README.md
# pysitcom


CHANGELOG.txt
Change Log
==========

0.1.0 (24/04/2021)
------------------
- First Release

test/test.py
import unittest
import numpy as np

from pysitcom import Comet, Sitcom


class TestComet(unittest.TestCase):

    def test_criteriaFail(self):
        with self.assertRaises(TypeError):
            Comet(["Bad", 0.5, 1])

    def test_criteriaType(self):
        with self.assertRaises(TypeError):
            Comet((0, 0.5, 1))

    def test_lackOfCriteria(self):
        with self.assertRaises(ValueError):
            Comet([])

    def test_noCoInRateCo(self):
        with self.assertRaises(ValueError):
            tmp = Comet([[2, 3]])
            tmp.rate_co()

    def test_preferenceMejFail(self):
        with self.assertRaises(ValueError):
            tmp = Comet([[0, 1]])
            tmp.generate_co()
            tmp.rate_co('mej')

    def test_preferenceValueFail(self):
        with self.assertRaises(ValueError):
            tmp = Comet([[0, 1]])
            tmp.generate_co()
            tmp.rate_co(value=8)

    def test_preferenceOptionFail(self):
        with self.assertRaises(ValueError):
            tmp = Comet([[0, 1]])
            tmp.generate_co()
            tmp.rate_co('bad')

    def test_noCoInRate(self):
        with self.assertRaises(ValueError):
            tmp = Comet([[2, 3]])
            tmp.rate(np.asarray([[0.2, 0.3]]))

    def test_noCoInChangePreferenceCo(self):
        with self.assertRaises(ValueError):
            tmp = Comet([[2, 3]])
            tmp.change_co_preference([0.1, 0.2])

    def test_highPreferenceValue(self):
        with self.assertRaises(ValueError):
            tmp = Comet([[2, 3]])
            tmp.generate_co()
            tmp.change_co_preference([0.1, 2])

    def test_lowPreferenceValue(self):
        with self.assertRaises(ValueError):
            tmp = Comet([[2, 3]])
            tmp.generate_co()
            tmp.change_co_preference([0.1, -1])

    def test_alternativesLen(self):
        with self.assertRaises(ValueError):
            tmp = Comet([[2, 3], [2, 3, 4]])
            tmp.generate_co()
            tmp.rate(np.array([[1, 2, 3]]))

    def test_alternativesValue(self):
        with self.assertRaises(ValueError):
            tmp = Comet([[1, 2], [2, 5, 8]])
            tmp.generate_co()
            tmp.rate(np.array([[4, 20], [2, -1]]))


class TestSitcom(unittest.TestCase):

    def test_alternativesValue(self):
        with self.assertRaises(ValueError):
            alternatives = np.array([[4, 20], [2, -1]])
            preference = [0.2, 0.4]
            criteria = [[1, 2], [9, 10]]
            Sitcom(alternatives, preference, criteria)

    def test_preferenceFail(self):
        with self.assertRaises(ValueError):
            alternatives = np.array([[1, 9], [2, 9]])
            preference = [1.2, 0.4]
            criteria = [[1, 2], [9, 10]]
            Sitcom(alternatives, preference, criteria)

    def test_criteriaFail(self):
        with self.assertRaises(TypeError):
            alternatives = np.array([[1, 9], [2, 9]])
            preference = [1.2, 0.4]
            criteria = [["False", 2], [9, 10]]
            Sitcom(alternatives, preference, criteria)


test/__init__.py
from test import *


pysitcom/Comet.py
import numpy as np
import itertools as it
import copy


class Comet:

    def __init__(self, criteria: list):
        """

        Parameters
        ----------
        criteria: list
            A list storing lists of criteria values
        """

        if not criteria:
            raise ValueError("Lack of criteria")

        if type(criteria) != list:
            raise TypeError("Wrong type of criteria: %s" % (repr(type(criteria))))

        for i, simpleCriterion in enumerate(criteria):
            for j, valueCriterion in enumerate(simpleCriterion):
                if type(valueCriterion) is not int and type(valueCriterion) is not float:
                    raise TypeError(
                        "Wrong type of %s value %s criterion: %s" % (repr(j), repr(i), repr(type(valueCriterion)))
                    )

        self._criteria: list = criteria
        self._sj: list = []
        self._co: np.ndarray = np.ndarray([0])

    def generate_co(self):
        """Generates characteristic object values based on the value of criteria"""

        criteria_iterator: list = list(it.product(*self._criteria))
        co: list = []

        for i, coord in enumerate(criteria_iterator):
            tmp: list = list(coord)
            tmp.append(0)
            co.append(tmp)

        self._co = np.asarray(co)

    def rate_co(self, preference: str = 'value', value: float = 0.0, mej: np.ndarray = np.ndarray([0])):
        """

        Parameters
        ----------
        preference: {'mej', 'random', 'value'}, optional
            Defines the value of preferences, default last when optional.
        value: int, optional
            Preference value that will be assigned to all characteristic objects (if preference is not 'mej' or
            'random'), default 0.0 value.
        mej: np.ndarray, optional
            MEJ matrix. Required when preference get 'mej' value.
        """

        if self._co.size == 0:
            raise ValueError(
                "Lack of characteristic objects"
            )

        if preference == 'mej':
            if mej.size == 0:
                raise ValueError(
                    "Lack of MEJ matrix"
                )
            else:
                p: list = self.__get_preference(mej)

        elif preference == 'random':
            p: list = list(np.random.rand(len(self._co)))

        elif preference == 'value':
            if value > 1 or value < 0:
                raise ValueError(
                    "Unexpected preference value: %s"
                    % (repr(value))
                )
            else:
                p: list = [value] * len(self._co)

        else:
            raise ValueError(
                "Unexpected preference option: %s"
                % (repr(preference))
            )

        self._co[:, -1] = p

    def change_co_preference(self, preference: list):
        """Changes the preference of characteristic objects

        Parameters
        ----------
        preference: list
            List of characteristic object preferences
        """
        if self._co.size == 0:
            raise ValueError(
                "Lack of characteristic objects"
            )

        if any(np.array(preference) > 1):
            raise ValueError(
                "Preference for a characteristic object too high"
            )

        if any(np.array(preference) < 0):
            raise ValueError(
                "Preference for a characteristic object too low"
            )

        self._co[:, -1] = preference

    def get_co_len(self):
        """Returns the number of characteristic objects

        Returns
        -------
            The number of characteristic objects
        """
        return len(self._co)

    def get_co_preference(self):
        """Returns preference values of characteristic objects

        Returns
        -------
            Preference values of characteristic objects
        """
        return self._co[:, -1]

    def rate(self, alternatives: np.ndarray) -> np.ndarray:
        """Calculates the value of preferences of given alternatives

        Parameters
        ----------
        alternatives: np.ndarray
            Numpy array of alternatives with their values for each criterion. Each alternative is a row in the array.

        Returns
        -------
            Numpy array of alternatives preferences
        """
        if self._co.size == 0:
            raise ValueError(
                "Lack of characteristic objects"
            )

        if not alternatives.shape[1] == self._co.shape[1] - 1:
            raise ValueError(
                "Incorrect number of criteria values for alternatives: %s" % (repr(alternatives.shape[0]))
            )

        for i in range(len(self._criteria)):
            if any(alternatives[:, i] < self._criteria[i][0]) or any(alternatives[:, i] > self._criteria[i][-1]):
                raise ValueError(
                    "The value of alternatives extends beyond bounds"
                )

        product: list = [[] for _ in range(len(alternatives))]
        preferences: np.ndarray = np.zeros(len(alternatives))
        alternatives: np.ndarray = copy.deepcopy(alternatives)

        for point in self._co:
            space_iter = self.__scope(point)

            for i, alternative in enumerate(alternatives):
                product[i] = [self.__tfn(coord, sp[0], sp[1], sp[2]) for sp, coord in zip(space_iter, alternative)]

            for i in range(len(alternatives)):
                preferences[i] += np.multiply(np.product(product[i]), point[-1])

            product = [[] for _ in range(len(alternatives))]

        return preferences

    def __get_preference(self, mej: np.ndarray) -> list:
        """Calculates preference for characteristic objects based on MEJ matrix"""

        self._sj: np.ndarray = np.zeros(len(mej))

        for i in range(len(self._sj)):
            self._sj[i] = sum(mej[i, :])

        sj_tmp: np.ndarray = copy.copy(self._sj)
        k: int = len(np.unique(self._sj))
        d: float = float(0)
        preference: np.ndarray = np.zeros(len(self._sj))

        for _ in range(k):
            index = np.where(sj_tmp == min(sj_tmp))
            preference[index] = float(d)
            d += 1 / (k - 1)
            sj_tmp[index] = max(sj_tmp) + 1

        return list(preference)

    def __scope(self, point: list) -> list:
        """Returns intervals in which the values of a characteristic object are contained

        Parameters
        ----------
        point: list
            Characteristic object

        Returns
        -------
            Characteristic space
        """
        scope_list: list = []
        for var, criterion in zip(point[:-1], self._criteria):
            ind: int = criterion.index(var)
            if ind == len(criterion) - 1:
                scope_list.append([criterion[ind - 1], criterion[ind], criterion[ind]])
            elif ind == 0:
                scope_list.append([criterion[ind], criterion[ind], criterion[ind + 1]])
            else:
                scope_list.append([criterion[ind - 1], criterion[ind], criterion[ind + 1]])

        return scope_list

    @staticmethod
    def __tfn(x: float, a: float, m: float, b: float) -> float:
        """Triangular Fuzzy Number method

        Parameters
        ----------
        x: float
            Representative crisp value
        a: float
            Smallest likely value
        m: float
            Most probable value
        b: float
            Largest possible value

        Returns
        -------
            Defuzzification value
        """
        if x < a or x > b:
            return float(0)

        elif a <= x < m:
            return (x-a) / (m-a)

        elif m < x <= b:
            return (b-x) / (b-m)

        elif x == m:
            return float(1)


pysitcom/similarity.py
import numpy as np


def ws(x: list, y: list) -> float:
    """Calculates the value of WS similarity coefficient
    Parameters
    ----------
    x: list
        List of values for similarity comparison
    y: list
        List of values for similarity comparison
    Returns
    -------
        Value of WS similarity coefficient
    """
    n = len(x)
    x = np.asarray(x)
    y = np.asarray(y)
    return 1 - sum(np.float_power(2, -x) * np.abs(x - y) / np.maximum(np.abs(1 - x), np.abs(n - x)))


def rw(x: list, y: list) -> float:
    """Calculates the value of the weighted Spearman correlation coefficient
    Parameters
    ----------
    x: list
        List of values for similarity comparison
    y: list
        List of values for similarity comparison
    Returns
    -------
         Value of Spearman weighted correlation coefficient
    """
    n = len(x)
    x = np.asarray(x)
    y = np.asarray(y)
    return 1 - 6 * sum(((x - y) ** 2) * (2 * n - x - y + 2)) / (n ** 4 + n ** 3 - n ** 2 - n)


pysitcom/Sitcom.py
from .Comet import Comet
import numpy as np
import copy
from mlrose import hill_climb, simulated_annealing, ContinuousOpt, CustomFitness
from pyswarms import single


# 	Stochastic IdenTifiCation Of Model
class Sitcom:

    def __init__(self, alternatives: np.ndarray, preference: list, criteria: list, stochastic_method: str = 'pso',
                 iterations: int = 1000):
        """

        Parameters
        ----------
        alternatives: np.ndarray
            Numpy array of alternatives with their values for each criterion. Each alternative is a row in the array.
        stochastic_method: {'hill-climbing', 'simulated-annealing', 'pso'}
            Defines a stochastic optimization method
        criteria: list
            A list storing lists of criteria values
        iterations: int, optional
            Number of iterations in the optimization method
        """

        for i in range(len(criteria)):
            if any(alternatives[:, i] < criteria[i][0]) or any(alternatives[:, i] > criteria[i][-1]):
                raise ValueError(
                    "The value of alternatives extends beyond bounds"
                )

        if any(np.array(preference) > 1):
            raise ValueError(
                "Preference for an alternatives too high"
            )

        if any(np.array(preference) < 0):
            raise ValueError(
                "Preference for an alternatives too low"
            )

        for i, simpleCriterion in enumerate(criteria):
            for j, valueCriterion in enumerate(simpleCriterion):
                if type(valueCriterion) is not int and type(valueCriterion) is not float:
                    raise TypeError(
                        "Wrong type of %s value %s criterion: %s" % (repr(j), repr(i), repr(type(valueCriterion)))
                    )

        self._alternatives: np.ndarray = alternatives
        self._alternativesPreference: list = preference
        self._stochasticMethod: str = stochastic_method
        self._criteria: list = criteria
        self._iterations: int = iterations

    def generate_criteria(self):
        """Creates criteria based on a alternatives"""

        if self._criteria:
            self._criteria = []

        n, _ = self._alternatives.shape

        for number in range(n):
            self._criteria.append([np.min(self._alternatives[:, number]), np.mean(self._alternatives[:, number]),
                                   np.max(self._alternatives[:, number])])

    def rate(self, co_type: str, co_value: float):
        """Calculates the preference of characteristic objects using the defined stochastic method.

        Parameters
        ----------
        co_type: str
            Type of preference values of objects characteristic of the COMET method
        co_value: float
            The value of preferences of objects characteristic of the COMET method
        Returns
        -------

        """

        model = Comet(self._criteria)
        model.generate_co()
        model.rate_co(co_type, co_value)

        dict_arg = {'model': model, 'alternatives': self._alternatives, 'preference': self._alternativesPreference}

        if self._stochasticMethod == "hill-climbing":
            problem = ContinuousOpt(model.get_co_len(), CustomFitness(self._mlrose_fitness, **dict_arg),
                                    maximize = False, step = 0.01)
            pos, _, cost_history = hill_climb(problem, max_iters = self._iterations, curve = True,
                                              init_state = model.get_co_preference())
            cost_history = np.abs(cost_history)

        elif self._stochasticMethod == "simulated-annealing":
            problem = ContinuousOpt(model.get_co_len(), CustomFitness(self._mlrose_fitness, **dict_arg),
                                    maximize = False, step = 0.01)
            pos, _, cost_history = simulated_annealing(problem, max_iters = self._iterations, curve = True,
                                                       init_state = model.get_co_preference())
            cost_history = np.abs(cost_history)

        elif self._stochasticMethod == "pso":
            bound_max = np.ones(model.get_co_len())
            bound_min = np.zeros(model.get_co_len())
            bounds = (bound_min, bound_max)
            options = {'c1': 0.5, 'c2': 0.3, 'w': 0.9}

            optimizer = single.GlobalBestPSO(n_particles = 20, dimensions = model.get_co_len(), options = options,
                                             bounds = bounds)
            cost, pos = optimizer.optimize(self._pso_fitness, self._iterations, **dict_arg)
            cost_history = optimizer.cost_history

        else:
            raise ValueError(
                "Wrong optimization method has been determined: %s"
                % (repr(self._stochasticMethod))
            )

        return pos, cost_history

    @staticmethod
    def _pso_fitness(x: np.ndarray, model: Comet, alternatives: np.ndarray, preference: list):
        """Fitness function for the PSO method

        Parameters
        ----------
        x: np.ndarray
            Position of the particles
        model: Comet
            Comet model with preference for characteristic objects
        alternatives: np.ndarray
            Numpy array of alternatives with their values for each criterion. Each alternative is a row in the array.
        preference:
            Reference preference for alternatives
        Returns
        -------
            Fitness function values for particles
        """

        fit_values = []
        for pref in x:
            model_tmp = copy.deepcopy(model)
            model_tmp.change_co_preference(pref)
            preference_tmp = model_tmp.rate(alternatives)
            fit_values.append(np.sum(np.abs(np.subtract(preference_tmp, preference))))
        return fit_values

    @staticmethod
    def _mlrose_fitness(x: np.ndarray, model: Comet, alternatives: np.ndarray, preference: np.ndarray):
        """Fitness function for the hill climbing and simulated annealing methods

        Parameters
        ----------
        x: np.ndarray
            Numpy array of preference values for characteristic objects
        model: Comet
            Comet model with preference for characteristic objects
        alternatives: np.ndarray
            Numpy array of alternatives with their values for each criterion. Each alternative is a row in the array.
        preference:
            Reference preference for alternatives
        Returns
        -------
            Fitness function value
        """

        model_tmp = copy.deepcopy(model)
        model_tmp.change_co_preference(list(x))
        preference_tmp = model_tmp.rate(alternatives)
        return np.sum(np.abs(np.subtract(preference_tmp, preference)))
